lib/SILOptimizer/Utils/Existential.cpp

```diff
@@ -311,3 +312,83 @@ ConcreteExistentialInfo::ConcreteExistentialInfo(Operand &openedUse) {
   }
   assert(isValid());
 }
+
+/// Derive a concrete type of self and conformance from protocol conformance
+/// analysis If successful, initializes a valid ConcreteExistentialInfo.
+ConcreteExistentialInfo::ConcreteExistentialInfo(
+    ProtocolConformanceAnalysis *PCA, ClassHierarchyAnalysis *CHA,
+    SILBuilder &Builder, ProtocolDecl *Protocol, FullApplySite Apply,
+    SILValue &Arg, Operand &ArgOperand) {
```

Comment:
This doesn't seem like a clean encapsulation anymore. Why can't the caller just pass in `ConcreteType` and `Protocol` when it can obtain those from its analyses? Why do you need to pass in `PCA` and `CHA`?

`Apply`, `Arg`, and `ArgOperand` all seem redundant. Why don't we just pass in a single `openedUse` (the open_existential_ref's use) just like the original constructor?

Also, the top-level comment on `struct ConcreteExistentialInfo` should be amended to mention new SIL patterns that we handle. Or move the existing comments onto the original constructor.

Reply:
The reason I wanted it this way is that I can create ConcreteExistentialInfo from another optimization https://github.com/apple/swift/pull/17366. I did not want code duplication later. May be for now, I will simplify the constructor as you suggested and worry about it later with a common interface for the other PR. Again, will update the comment.